src/core/lib/security/security_connector/local/local_security_connector.cc

```diff
@@ -70,39 +70,19 @@ void local_check_peer(tsi_peer peer, grpc_endpoint* ep,
                       grpc_core::RefCountedPtr<grpc_auth_context>* auth_context,
                       grpc_closure* on_peer_checked,
                       grpc_local_connect_type type) {
-  int fd = grpc_endpoint_get_fd(ep);
-  grpc_resolved_address resolved_addr;
-  memset(&resolved_addr, 0, sizeof(resolved_addr));
-  resolved_addr.len = GRPC_MAX_SOCKADDR_SIZE;
   bool is_endpoint_local = false;
-  if (getsockname(fd, reinterpret_cast<grpc_sockaddr*>(resolved_addr.addr),
-                  &resolved_addr.len) == 0) {
-    grpc_resolved_address addr_normalized;
-    grpc_resolved_address* addr =
-        grpc_sockaddr_is_v4mapped(&resolved_addr, &addr_normalized)
-            ? &addr_normalized
-            : &resolved_addr;
-    grpc_sockaddr* sock_addr = reinterpret_cast<grpc_sockaddr*>(&addr->addr);
-    // UDS
-    if (type == UDS && grpc_is_unix_socket(addr)) {
+
+  char* peer_address = ep->vtable->get_peer(ep);
+  if (type == UDS && strncmp("uds", peer_address, 3)) {
+    is_endpoint_local = true;
+  } else if (type == LOCAL_TCP) {
+    if (strncmp("ipv4:127.0.0.1", peer_address, 14)) {
+      is_endpoint_local = true;
+    } else if (strncmp("ipv6:[::1]", peer_address, 10)) {
```

Comment:
I am a little hesitant to use these hard-coded string comparison to determine whether the endpoint is a local or not. If we ever change the implementation of `get_peer` (e.g., using a new format) in future, this logic will fail (which is not the case using the original implementation BTW).

Reply:
Instead, can we use `get_peer()` method only when `fd` is not available? i.e., when `tcp_custom` is used?

Reply:
Do you have suggestions that where I can find constants that represent those values? The struct looks like defined by gRPC. I couldn't find a way to convert a string to `grpc_resolved_address`, except invoking name resolver again. That seems sub-optimal.

Reply:
Can you explain more about the value of having two versions of check? If anyone in Core team wants to modify this logic, the test will fail. I can lift those values to be constants in the namespace. WDYT?

Reply:
https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/sockaddr_utils.cc#L219 is the place that bake `peer_address`.

Reply:
It seems more reasonable to evaluate fd using syscalls to determine the socket property. And if `fd` is not available, then we can only do our best to make the decision.

Reply:
I agree with @yihuazhang here -- it does not make sense to use anything other than the fd here.

I think this problem is a symptom of the fact that the asyncio code was implemented as a custom iomgr instead of via the EventManager interface.  The custom iomgr interface is an incomplete abstraction, and the fact that it pretends to support endpoints but its endpoints don't actually provide the full endpoint API (which includes access to the underlying fd) causes problems exactly like this one.  This is one of the many reasons why I want to get rid of the iomgr API.  If asyncio was implemented via the EventManager interface, then I don't think this would be a problem.

Given that asyncio is currently implemented the wrong way, if you want to add some sort of fallback in local credentials for the case where the fd is not available, that's fine as a temporary solution.  But we should not stop using the fd in cases that are not broken this way.  And we should definitely be actively working toward an EventManager-based approach for asyncio to make this kind of temporary hack go away.